lib/hooks/index.js

```diff
@@ -97,25 +58,45 @@ function findModuleVersion(request, parent, load) {
     if (mainScriptDir.indexOf(paths[i]) === 0) {
       // This is unsafe on windows if request contains a namespace
       // TODO: convert path separators in request
-      var pjson = load(path.join(paths[i], request, 'package.json'));
-      return pjson.version;
+      return path.join(paths[i], request);
     }
   }
   return null;
 }
 
+function findModuleVersion(modulePath, load) {
+  return modulePath ? load(path.join(modulePath, 'package.json')).version :
+    process.version;
+}
+
 function activate(agent) {
 
   logger = agent.logger;
 
-  function loadHookAndPatch(loadFn, instrumentation) {
-    assert(instrumentation.module);
-    assert(!instrumentation.hook);
-
+  function loadHookAndPatch(loadFn, instrumentation, moduleRoot, version) {
+    var versionHooks = instrumentation.versions[version];
     // Load the hook. This file, i.e. index.js, becomes the parent module.
-    var moduleHook = loadFn(instrumentation.file, module, false);
-    instrumentation.hook = moduleHook;
-    moduleHook.patch(instrumentation.module, agent, instrumentation.version);
+    if (!versionHooks) {
+      var moduleHook = loadFn(instrumentation.file, module, false);
+      versionHooks = moduleHook(version, agent);
+    }
+    Object.keys(versionHooks).forEach(function(file) {
```

Comment:
It is the set of hooks that should be applied for the given version where a hook is a file -> patch/unpatch map. Is there a name that would clarify this structure better?